ocaml/xapi/stream_vdi.ml

```diff
@@ -407,18 +408,26 @@ let recv_all_vdi refresh_session ifd (__context:Context.t) rpc session_id ~has_i
              in
              Unixext.really_read ifd buffer 0 (Int64.to_int length);
              Unix.write ofd buffer 0 (Int64.to_int length) |> ignore;
-             let csum = Sha1.to_hex (Sha1.string (Bytes.unsafe_to_string buffer)) in
+             
+             let buffer_string = Bytes.unsafe_to_string buffer in
+             
+             let csum_hdr = Tar_unix.Header.get_next_header ifd in (* Header of the checksum file *)
+	     let csum_file_name = csum_hdr.Tar_unix.Header.file_name in
+             
+             let csum = if Filename.check_suffix csum_file_name checksum_extension (* Infer the hash algorithm *)
+             	then Printf.sprintf "%016LX" (XXH64.hash buffer_string)
+             	else Sha1.to_hex (Sha1.string buffer_string)
```

Comment:
This works, although another way to implement this would be to have a function that receives the `Filename.extension csum_file_name`, and the string buffer, does a `match` on the extension and returns the checksum.